Data/ByteString/Internal/Type.hs

```diff
@@ -641,10 +643,13 @@ createFpUptoN' l action = do
 --
 createFpAndTrim :: Int -> (ForeignPtr Word8 -> IO Int) -> IO ByteString
 createFpAndTrim l action = do
-    fp <- mallocByteString l
-    l' <- action fp
-    if assert (0 <= l' && l' <= l) $ l' >= l
-        then return $! BS fp l
+  fp <- mallocByteString l
+  l' <- action fp
+  if assert (0 <= l' && l' <= l) $ l' >= l
+    then return $! BS fp l
+    else
+      if l < 4096
```

Comment:
This condition is the main tricky bit.

* Large `MutableByteArray#`s do not return their memory to the RTS when we shrink them in-place. But they do return their memory to the RTS when we make a copy.
* Small pinned `MutableByteArray#`s may well retain all of their memory regardless of whether we shrink in place or create a copy, because of how they are currently allocated in GHC. (There are probably several GHC issues related to this topic. See for example [this one](https://gitlab.haskell.org/ghc/ghc/-/issues/22768) and [this one](https://gitlab.haskell.org/ghc/ghc/-/issues/19175). It's not clear if or when this will be improved.)

So we want to make a copy when:
* We expect the underlying buffer to be a large heap object, and
* the amount of space shrinking would leak is large enough to care about.

If I recall correctly, the exact threshold for when a buffer is allocated as a large heap object depends on both the platform and the GHC version, but has been somewhere around 3K for 64-bit systems and half that for 32-bit systems for some time now.

(This reasoning should be documented in the code.)